File: 105-construct-binary-tree-from-preorder-and-inorder-traversal/construct-binary-tree-from-preorder-and-inorder-traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
    unordered_map<int,int> inorderIdx;
    int pre_idx = 0;

public:
    TreeNode* buildTree(vector<int>& preorder, vector<int>& inorder) {
        for(int i=0; i<inorder.size(); i++){
            inorderIdx[inorder[i]] = i;
        }

        return dfs(preorder, 0, inorder.size()-1);
    }

    TreeNode* dfs(vector<int>& preorder, int left, int right){
        if(left > right) return nullptr;

        TreeNode* root = new TreeNode(preorder[pre_idx++]);
        int mid = inorderIdx[root->val];
        root->left = dfs(preorder,left, mid-1);
        root->right = dfs(preorder,mid+1, right);

        return root;
    }
};

File: 105-construct-binary-tree-from-preorder-and-inorder-traversal/construct-binary-tree-from-preorder-and-inorder-traversal.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {number[]} preorder
 * @param {number[]} inorder
 * @return {TreeNode}
 */
var buildTree = function(preorder, inorder) {
   let inorderMap = new Map();
   let preIdx = 0;

   for(let i=0; i<inorder.length; i++){
    inorderMap.set(inorder[i], i)
   }
   
   const treeHelper = (s, e) => {
    if(s > e) return null;

    let root = new TreeNode(preorder[preIdx]);
    preIdx++;

    let inorderIdx = inorderMap.get(root.val); // root idx

    root.left = treeHelper(s, inorderIdx-1);
    root.right = treeHelper(inorderIdx+1, e);

    return root;
   }

   return treeHelper(0, inorder.length-1)
};